pinot-spi/src/main/java/org/apache/pinot/spi/trace/Tracing.java

```diff
@@ -332,17 +333,27 @@ public static boolean isInterrupted() {
           || Tracing.getThreadAccountant().isQueryTerminated();
     }
 
-    public static void sampleAndCheckInterruption() {
-      if (isInterrupted()) {
+    public static void checkInterruptionAndThrow(ThreadResourceUsageAccountant accountant) {
+      if (Thread.interrupted()) {
         throw new EarlyTerminationException("Interrupted while merging records");
       }
+
+      if (accountant.isAnchorThreadInterrupted() || accountant.isQueryTerminated()) {
```

Comment:
Is this guaranteed to be OOM protection? Is regular query cancellation also hitting this condition?

Reply:
Yes. Anchor thread is interrupted by WatcherTask only